numpy/core/arrayprint.py

```diff
@@ -503,18 +518,25 @@ def array2string(a, max_line_width=None, precision=None,
     '[0x0L 0x1L 0x2L]'
 
     """
-    # Deprecation 05-16-2017  v1.14
-    if style is not np._NoValue:
-        warnings.warn("'style' argument is deprecated and no longer functional",
-                      DeprecationWarning, stacklevel=3)
+    legacy = kwarg.pop('legacy', None)
+    if kwarg:
+        msg = "array2string() got unexpected keyword argument '{}'"
+        raise TypeError(msg.format(kwarg.popitem()[0]))
 
     overrides = _make_options_dict(precision, threshold, edgeitems,
                                    max_line_width, suppress_small, None, None,
-                                   sign, formatter)
+                                   sign, formatter, legacy)
     options = _format_options.copy()
     options.update(overrides)
 
-    if a.size == 0:
+    if options['legacy'] and a.shape == () and not a.dtype.names:
+        if style is None:
+            style = repr
+        return style(a.item())
```

Comment:
I think I'd prefer that the `style` argument remain deprecated _unless_ `legacy=True`.

Reply:
That seems like a good idea.

Reply:
If you write that up in #9201 I will consider it, though I still slightly disagree since it seems like an almost useless back-compatibility break.

Also, at first glance it's unclear how to do that, since to get rid of `style` you need to hardcode the 0d `str` so it is unaffected by `np.set_printoptions`... but then `array2string` would never get called, nor its legacy branch. But probably there is a way.